### PR TITLE
feat(sparsity): Add FisherPruner — FIM-guided weight pruning with calibration-based eFIM accumulation

### PR DESCRIPTION
**Pull request by ramkrishs**
## Summary

This PR introduces `FisherPruner`, a new `BaseSparsifier` subclass that prunes neural network weights using the diagonal of the empirical Fisher Information Matrix (eFIM) rather than weight magnitude alone or the Wanda activation-norm criterion.

### Motivation

Existing sparsifiers in torchao use either magnitude (`WeightNormSparsifier`) or weight × activation-norm (`WandaSparsifier`) as the pruning criterion. The Fisher Information Matrix provides a principled, loss-aware alternative: a parameter's eFIM diagonal entry approximates the expected curvature of the loss w.r.t. that parameter. Pruning low-eFIM weights minimises the expected increase in loss, a well-known result from Optimal Brain Damage (LeCun et al., 1990) and Optimal Brain Surgeon (Hassibi & Stork, 1993).

### Algorithm

The diagonal eFIM is approximated empirically as the mean squared gradient across calibration batches:

$$F_{ii} \approx \frac{1}{T} \sum_{t=1}^{T} \left(\frac{\partial \ell_t}{\partial \theta_i}\right)^2$$

Weights with the *lowest* eFIM scores are pruned first — their removal has the smallest expected impact on the loss.

### API (mirrors WandaSparsifier)

```python
from torchao.sparsity import FisherPruner

pruner = FisherPruner(sparsity_level=0.5)
pruner.prepare(model, config=None)

# Calibration: accumulate FIM statistics over representative data
for X, y in calibration_loader:
    loss = criterion(model(X), y)
    loss.backward()
    pruner.accumulate_fim()   # accumulate squared gradients
    model.zero_grad()

pruner.step()         # apply FIM-guided masks
pruner.squash_mask()  # finalise: remove parametrizations
```

Supports:
- **Unstructured sparsity** — arbitrary `sparsity_level` (0–1)
- **Semi-structured 2:N sparsity** — via `semi_structured_block_size`
- **Per-layer config** — `config=[{"tensor_fqn": "layer.weight"}]`
- **Graceful fallback** — warns and falls back to magnitude pruning if no calibration data is provided

### Files changed

| File | Change |
|------|--------|
| `torchao/sparsity/fisher_pruner.py` | New `FisherPruner` class (≈220 lines) |
| `torchao/sparsity/__init__.py` | Export `FisherPruner` |
| `test/sparsity/test_fisher_pruner.py` | 14 unit tests |

### Tests

```
Ran 14 tests in 0.070s — OK
```

Covers: construction validation, `prepare` parametrization, `accumulate_fim` accumulation, no-gradient safety, `squash_mask` cleanup, unstructured sparsity level correctness, known-weight pruning direction, 2:4 semi-structured sparsity, fallback-to-magnitude warning, and per-layer custom config.

### References

- LeCun et al., *Optimal Brain Damage*, NeurIPS 1990
- Hassibi & Stork, *Optimal Brain Surgeon*, NeurIPS 1993
- Singh & Alistarh, *WoodFisher: Efficient Second-Order Approximation for Neural Network Compression*, NeurIPS 2020

---

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD 3-Clause License.

### CODE REVIEW

**Review by jerryzh168 — COMMENTED** [review]
can you put this in https://github.com/pytorch/ao/tree/main/torchao/prototype/sparsity

**Comment by jerryzh168** [comment]
also please create a readme to talk about things in summary + show some e2e model result (accuracy impact) on a popular model

**Comment by ramkrishs** [comment]
Thanks @jerryzh168 — moved to `torchao/prototype/sparsity/sparsifier/` as requested.

**Changes in latest commit (`aeb70b66`):**
- `torchao/prototype/sparsity/sparsifier/fisher_pruner.py` — new location
- `torchao/prototype/sparsity/__init__.py` — exports `FisherPruner`
- `torchao/sparsity/__init__.py` — removed (was premature for stable)
- `test/prototype/sparsity/test_fisher_pruner.py` — test moved to match

New import path:
```python
from torchao.prototype.sparsity import FisherPruner
```

**Comment by ramkrishs** [comment]
Sure i am collecting some results will publish and update here.
